firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigSharedPrefsClient.java

```diff
@@ -234,21 +252,91 @@ void setLastTemplateVersion(long templateVersion) {
   BackoffMetadata getBackoffMetadata() {
     synchronized (backoffMetadataLock) {
       return new BackoffMetadata(
-          frcMetadata.getInt(NUM_FAILED_FETCHES_KEY, NO_FAILED_FETCHES),
-          new Date(frcMetadata.getLong(BACKOFF_END_TIME_IN_MILLIS_KEY, NO_BACKOFF_TIME_IN_MILLIS)));
+          frcSharedPrefs.getInt(NUM_FAILED_FETCHES_KEY, NO_FAILED_FETCHES),
+          new Date(
+              frcSharedPrefs.getLong(BACKOFF_END_TIME_IN_MILLIS_KEY, NO_BACKOFF_TIME_IN_MILLIS)));
     }
   }
 
   void setBackoffMetadata(int numFailedFetches, Date backoffEndTime) {
     synchronized (backoffMetadataLock) {
-      frcMetadata
+      frcSharedPrefs
           .edit()
           .putInt(NUM_FAILED_FETCHES_KEY, numFailedFetches)
           .putLong(BACKOFF_END_TIME_IN_MILLIS_KEY, backoffEndTime.getTime())
           .apply();
     }
   }
 
+  public void setCustomSignals(Map<String, String> newCustomSignals) {
+    synchronized (customSignalsLock) {
```

Comment:
Synchronization ensures thread safety and prevents race conditions when accessing and modifying custom signals. This is crucial because multiple threads might access this code concurrently. The synchronized block acts as a lock, allowing only one thread at a time to access and modify the signals.

We use multiple synchronized blocks with different locks for unrelated data within shared preferences for Remote Config. This approach allows for concurrent access to different parts of the configuration data while still ensuring thread safety for each individual piece.

For a more detailed internal discussion, please refer this: [Link](https://docs.google.com/document/d/1Ittbh_FTxRjsSovB9IVVOrmkOg_nQON3QWOsP_UrcxE/edit?disco=AAABWhm7WvA)